// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// =============================================================================
syntax = "proto3";

package tfx_bsl.sketches;

// InputType represents the type of values consumed by a sketch. Values of
// distinct type that are encoded and decoded identically may be assigned
// the same InputType.
message InputType {
enum Type {
  UNSET = 0;
  // Represents a value stored as a raw string (identity decoder).
  RAW_STRING = 1;
  // Represents a float or double, stored internally in hex exponential format.
  FLOAT = 2;
}
}

// Specifies how the KMV sketch is formatted for serialization.
message Kmv {
  // Max number of values in the sketch.
  int32 num_buckets = 1;
  // Values of the hashes.
  repeated fixed64 hashes = 2 [packed = true];
  // Largest hash value in the sketch.
  fixed64 max_limit = 3;
  InputType.Type input_type = 4;
}

// Specifies how the MisraGries sketch is formatted for serialization.
message MisraGries {
  // Max number of values in the sketch.
  int32 num_buckets = 1;
  // Names of items.
  repeated bytes items = 2;
  // Weights or counts of items.
  repeated double weights = 3 [packed = true];
  // Extra items with zero weights.
  repeated bytes extra_items = 9;
  // The quantity decremented from the stored weights.
  double delta = 4;

  // If true, the sketch will replace invalid utf8 values added to it with
  // `invalid_utf8_placeholder`.
  bool replace_invalid_utf8_values = 5;
  string invalid_utf8_placeholder = 6;
  // If non-negative, the sketch will replace string values longer than
  // the given threshold with `large_string_placeholder`.
  int32 large_string_threshold = 7;
  string large_string_placeholder = 8;
  InputType.Type input_type = 11;
}

// Specifies how the Quantiles Sketch is formatted for serialization.
message Quantiles {
  // Quantiles computation stream. Contains multiple levels of summaries and a
  // buffer of unprocessed elements.
  message Stream {
    // Contains processed input elements.
    message Summary {
      repeated double value = 1 [packed = true];
      repeated double weight = 2 [packed = true];
      repeated double min_rank = 3 [packed = true];
      repeated double max_rank = 4 [packed = true];
    }
    repeated Summary summaries = 1;
    // Contains unprocessed input elements.
    message Buffer {
      repeated double value = 1 [packed = true];
      repeated double weight = 2 [packed = true];
    }
    Buffer buffer = 2;
    // Parameters of the streams producing the sketch.
    // Two streams with different parameters may be merged, but the error bounds
    // may change in that case.
    double eps = 3;
    int64 max_num_elements = 4;
    // Indicates whether the stream was compacted before serialization.
    bool compacted = 5;
  }
  // Two sketches may only be merged on stream-to-stream basis.
  repeated Stream streams = 1;
}
